File: translator.js
const CODON_TABLE = {
  'AUG': { name: 'Methionine (Start)', details: 'Initiates protein synthesis.' },
  'UUU': { name: 'Phenylalanine', details: 'An essential amino acid used in protein biosynthesis.' },
  'UUC': { name: 'Phenylalanine', details: 'An essential amino acid used in protein biosynthesis.' },
  'UUA': { name: 'Leucine', details: 'Helps with protein structure and energy production.' },
  'UUG': { name: 'Leucine', details: 'Helps with protein structure and energy production.' },
  'CUU': { name: 'Leucine', details: 'Helps with protein structure and energy production.' },
  'CUC': { name: 'Leucine', details: 'Helps with protein structure and energy production.' },
  'CUA': { name: 'Leucine', details: 'Helps with protein structure and energy production.' },
  'CUG': { name: 'Leucine', details: 'Helps with protein structure and energy production.' },
  'AUU': { name: 'Isoleucine', details: 'Supports muscle metabolism and immune function.' },
  'AUC': { name: 'Isoleucine', details: 'Supports muscle metabolism and immune function.' },
  'AUA': { name: 'Isoleucine', details: 'Supports muscle metabolism and immune function.' },
  'GUU': { name: 'Valine', details: 'Essential for muscle growth and tissue repair.' },
  'GUC': { name: 'Valine', details: 'Essential for muscle growth and tissue repair.' },
  'GUA': { name: 'Valine', details: 'Essential for muscle growth and tissue repair.' },
  'GUG': { name: 'Valine', details: 'Essential for muscle growth and tissue repair.' },
  'UCU': { name: 'Serine', details: 'Plays a role in metabolism and immune response.' },
  'UCC': { name: 'Serine', details: 'Plays a role in metabolism and immune response.' },
  'UCA': { name: 'Serine', details: 'Plays a role in metabolism and immune response.' },
  'UCG': { name: 'Serine', details: 'Plays a role in metabolism and immune response.' },
  'UGA': { name: 'STOP', details: '' },
  // Add similar details for other codons...
};

function translateCodons() {
  const dna = document.getElementById("dnaInput").value.trim().toUpperCase().replace(/[^ATCG]/g, '');
  const rna = dna.replace(/T/g, 'U');
  const output = document.getElementById("output");
  const proteinCountElement = document.getElementById("proteinCount");
  const chromosomeCountElement = document.getElementById("chromosomeCount");

  if (rna.length < 3) {
    output.innerText = "Please enter a valid DNA sequence with at least one codon.";
    proteinCountElement.innerText = 0;
    chromosomeCountElement.innerText = 0;
    return;
  }

  let translation = '';
  let proteinCount = 0;
  let chromosomeCount = Math.ceil(dna.length / 1000000); // Assuming 1 chromosome per 1M base pairs

  for (let i = 0; i <= rna.length - 3; i += 3) {
    let codon = rna.substring(i, i + 3);
    let proteinData = CODON_TABLE[codon];
    if (!proteinData) {
      translation += `<span class="codon">${codon}</span> → <span class="protein">Unknown</span>\n`;
      continue;
    }
    if (proteinData.name === 'STOP') break;
    translation += `<span class="codon">${codon}</span> → <span class="protein">${proteinData.name}</span>\n`;
    translation += `<span class="details">${proteinData.details}</span>\n\n`;
    proteinCount++;
  }

  output.innerHTML = translation || "No valid codons found.";
  proteinCountElement.innerText = proteinCount;
  chromosomeCountElement.innerText = chromosomeCount;
}

function copyOutput() {
  const tempDiv = document.createElement("div");
  tempDiv.innerHTML = document.getElementById("output").innerHTML;

  const spans = tempDiv.querySelectorAll('span');
  spans.forEach(span => {
    const textNode = document.createTextNode(span.innerText);
    span.parentNode.replaceChild(textNode, span);
  });

  const plainText = tempDiv.innerText;

  navigator.clipboard.writeText(plainText).then(() => {
    alert("Protein sequence copied to clipboard!");
  }).catch(err => {
    alert("Failed to copy. Please try again.");
    console.error(err);
  });
}

File: translator.py
import FreeSimpleGUI as sg
import pyperclip

CODON_TABLE = {
    'AUG': 'Methionine (Start)',
    'UUU': 'Phenylalanine', 'UUC': 'Phenylalanine',
    'UUA': 'Leucine', 'UUG': 'Leucine',
    'CUU': 'Leucine', 'CUC': 'Leucine', 'CUA': 'Leucine', 'CUG': 'Leucine',
    'AUU': 'Isoleucine', 'AUC': 'Isoleucine', 'AUA': 'Isoleucine',
    'GUU': 'Valine', 'GUC': 'Valine', 'GUA': 'Valine', 'GUG': 'Valine',
    'UCU': 'Serine', 'UCC': 'Serine', 'UCA': 'Serine', 'UCG': 'Serine',
    'CCU': 'Proline', 'CCC': 'Proline', 'CCA': 'Proline', 'CCG': 'Proline',
    'ACU': 'Threonine', 'ACC': 'Threonine', 'ACA': 'Threonine', 'ACG': 'Threonine',
    'GCU': 'Alanine', 'GCC': 'Alanine', 'GCA': 'Alanine', 'GCG': 'Alanine',
    'UAU': 'Tyrosine', 'UAC': 'Tyrosine',
    'CAU': 'Histidine', 'CAC': 'Histidine',
    'CAA': 'Glutamine', 'CAG': 'Glutamine',
    'AAU': 'Asparagine', 'AAC': 'Asparagine',
    'AAA': 'Lysine', 'AAG': 'Lysine',
    'GAU': 'Aspartic acid', 'GAC': 'Aspartic acid',
    'GAA': 'Glutamic acid', 'GAG': 'Glutamic acid',
    'UGU': 'Cysteine', 'UGC': 'Cysteine',
    'UGG': 'Tryptophan',
    'CGU': 'Arginine', 'CGC': 'Arginine', 'CGA': 'Arginine', 'CGG': 'Arginine',
    'AGU': 'Serine', 'AGC': 'Serine',
    'AGA': 'Arginine', 'AGG': 'Arginine',
    'GGU': 'Glycine', 'GGC': 'Glycine', 'GGA': 'Glycine', 'GGG': 'Glycine',
    'UAA': 'STOP', 'UAG': 'STOP', 'UGA': 'STOP'
}

def translate_dna(sequence):
    sequence = sequence.upper().replace('T', 'U')
    result = []
    for i in range(0, len(sequence) - 2, 3):
        codon = sequence[i:i+3]
        amino = CODON_TABLE.get(codon, 'Unknown')
        if amino == 'STOP':
            break
        result.append(f"{codon} → {amino}")
    return '\n'.join(result)

layout = [
    [sg.Text("Enter DNA Sequence:")],
    [sg.Multiline(size=(60, 5), key='-INPUT-', font=("Courier", 12))],
    [sg.Button("Translate"), sg.Button("Clear"), sg.Button("Copy"), sg.Button("Exit")],
    [sg.Text("Protein Sequence:")],
    [sg.Multiline(size=(60, 15), key='-OUTPUT-', font=("Courier", 12), disabled=True)]
]

window = sg.Window("Codon Translator", layout, resizable=True)

while True:
    event, values = window.read()
    if event in (sg.WINDOW_CLOSED, 'Exit'):
        break
    elif event == 'Translate':
        output = translate_dna(values['-INPUT-'])
        window['-OUTPUT-'].update(output)
    elif event == 'Clear':
        window['-INPUT-'].update('')
        window['-OUTPUT-'].update('')
    elif event == 'Copy':
        pyperclip.copy(values['-OUTPUT-'])

window.close()
